dGame/dCinema/Recorder.h

```diff
@@ -0,0 +1,401 @@
+#pragma once
+
+#include "Player.h"
+#include "Game.h"
+#include "EntityManager.h"
+#include "tinyxml2.h"
+#include "Play.h"
+
+#include <chrono>
+
+namespace Cinema::Recording
+{
+
+class Record
+{
+public:
+	virtual void Act(Entity* actor) = 0;
+
+	virtual void Serialize(tinyxml2::XMLDocument& document, tinyxml2::XMLElement* parent) = 0;
+
+	virtual void Deserialize(tinyxml2::XMLElement* element) = 0;
+
+	float m_Delay;
+
+	std::string m_Name;
+};
+
+class Recorder
+{
+public:
+	Recorder();
+
+	~Recorder();
+
+	void AddRecord(Record* record);
+
+	void Act(Entity* actor, Play* variables = nullptr);
+
+	Entity* ActFor(Entity* actorTemplate, Entity* player, Play* variables = nullptr);
+
+	void StopActingFor(Entity* actor, Entity* actorTemplate, LWOOBJID playerID);
+
+	bool IsRecording() const;
+
+	void SaveToFile(const std::string& filename);
+
+	float GetDuration() const;
+	
+	static Recorder* LoadFromFile(const std::string& filename);
+
+	static void AddRecording(LWOOBJID actorID, Recorder* recorder);
+
+	static void StartRecording(LWOOBJID actorID);
+
+	static void StopRecording(LWOOBJID actorID);
+
+	static Recorder* GetRecorder(LWOOBJID actorID);
+
+	static void RegisterEffectForActor(LWOOBJID actorID, const int32_t& effectId);
+
+	static void LoadRecords(tinyxml2::XMLElement* root, std::vector<Record*>& records);
+
+	static void ActingDispatch(Entity* actor, const std::vector<Record*>& records, size_t index, Play* variables);
+
+private:
+	static void PlayerProximityDispatch(Entity* actor, const std::vector<Record*>& records, size_t index, Play* variables, std::shared_ptr<bool> actionTaken);
+
+	static void PathFindDispatch(Entity* actor, const std::vector<Record*>& records, size_t index, Play* variables);
```

Comment:
These could be moved to the translation unit for this file since they are static private members of a class, they dont really need to be in the header for it, what do you think?